### ISSUE TITLE
[css-text-3] text-transform: full-size-kana

### ISSUE THREAD

**Issue by frivoal** [w3c/csswg-drafts#3143]
Many years ago, we [discussed, and eventually rule against](https://www.w3.org/Search/Mail/Public/search?keywords=full-size-kana&hdr-1-name=subject&hdr-1-query=&index-grp=Public_FULL&index-type=t&type-index=www-style), a `full-size-kana` value to `text-transform`

It would have turned something like しゃ into しや. The two are not equivalent, (the first is pronounced “sha”, the second “shiya”), but at very small font sizes inside ruby text, this substitution is sometimes used to keep things at a legible size.

Without this text transform, authors wanting this effect will just hard code the transformed string into the source, which does the wrong thing from searches or speech synthesis.

The main argument against this feature was that this was a rather niche i18n feature, and if we started to add those to text-transform, there'd be no end to it, so it would be better to pursue a generic extension mechanism. I was one of those pushing this line of thought, and even drafted a possible generic mechanism: https://specs.rivoal.net/css-custom-tt/

I repent, and I think we should have added/kept `full-size-kana`.

* In the years since, there hasn't been an avalanche of requests for more i18n transforms like this one, so the slippery slope concern does seem to have been founded.
* In the meanwhile, the i18n/a11y problem that this was meant to solve remains unsolved.
* The custom text-transform mechanism has some interesting difficulties, and has not received any attention since then.

I think we should consider adding this value back.

**Comment by fantasai** [comment]
This was originally defined in CSS Text Level 3: https://www.w3.org/TR/2012/WD-css3-text-20120119/#text-transform
and was removed in 2012

**Comment by momdo** [comment]
FYI, see also [section 3.3.3 in jlreq](https://www.w3.org/TR/2012/NOTE-jlreq-20120403/#choice_of_size_for_ruby_characters):
> (note 3) | Slightly different from the question of the size of ruby, there is a question of whether or not small kana (cl-11) characters should be used in ruby annotations. Because the size is so small, there was no choice of using small kana in letterpress printing. Today there are cases where small kana are used in ruby annotations, but taking legibility into consideration, these cases should be limited to those where accurate readings are needed, such as for proper nouns.

**Comment by kojiishi** [comment]
> The main argument against this feature was that...

Wasn't the main argument about preferring [`font-variant-east-asian: ruby`](https://drafts.csswg.org/css-fonts-3/#ruby)?

**Comment by Crissov** [comment]
For the custom transforms mechanism I recently filed #3132, by the way.

**Comment by frivoal** [comment]
> Wasn't the main argument about preferring font-variant-east-asian: ruby?

I don't think it was the main argument back then, but it is certainly worth considering now.

`font-variant-east-asian: ruby` seems aimed at a slightly different problem: both our spec and the open-type spec seem to advocate using in on regular kana to get better visual pairing between the base text and the ruby text. (css-text does that by showing an example of the word しんかんせん, which does not contain small kana, with thicker glyphs, and [open type documentation](https://docs.microsoft.com/en-us/typography/opentype/spec/features_pt#tag-ruby) suggests using it on U+3042, which is not a small kana).

That said, it is true that it could also be used to show large glyphs for small kana. Should it:
* pro: It allows, on a per-font basis, to pick a different way to cope with overly small kana
* pro: It exists, so there's no need to introduce anything new in css
* con: Doing it with a text-transform works even if the font fails to download
* con: Doing it with a text-transform works even if there's no available font that uses this open-type feature that way

**Comment by upsuper** [comment]
I would also think it's better be addressed by `font-variant-east-asian: ruby`.

> con: Doing it with a text-transform works even if the font fails to download

I don't see why this is a con. Fonts provided by operating system is probably more likely to have better support on this kind of OpenType features. At least, Hiragino on macOS and Meiryo on Windows both have so (although Gecko has to avoid applying this feature on Windows due to a bug in old versions of Meiryo). None of them make the small kana's bigger, though.

**Comment by frivoal** [comment]
> Fonts provided by operating system is probably more likely to have better support on this kind of OpenType features.
> [...]
> None of them make the small kana's bigger, though.

System fonts could do it, but don't. In-the-wild fonts might do it, but I don't know any that does, and even if we find some, it will have to be used as a web font.

So, if an author wants this effect, they have to rely on a web font (which might fail to download), and they have a very limited choice of fonts which actually do this, given that this is not what the feature was advertised to be about in the open type spec, and that therefore most font don't do the small-kana -> big-kana conversion.

That makes me think that the text transform avenue is still quite a reasonable one.

**Comment by fantasai** [comment]
I think the more important reason not to rely on `font-variant: ruby` is that `font-variant: ruby` is primarily about matching the weight of the glyph: whether you want to also use full-size kana is a separate question, you might or you might not. As has been noted, using full-size kana makes the kana a little easier to read because it's bigger, but it loses the distinction between two letters which are in fact different letters. It would be better if fonts were not encouraged to use full-size kana when `font-variant: ruby` is on so that the choice about whether full-size the kana remains open to the author and user.

**Comment by macnmm** [comment]
I actually am not in favor of this feature for ruby -- when the ruby is over the base text, it is not correct to use small kana glyphs, or glyphs of the same weight. The glyphs for small kana with ruby variants should be full size and thicker. My reasoning is about the text stream and placement being separable -- if you place the ruby after the text in parentheses, you want small kana, but over the base text you do not. Transforming the text stream would be an extra destructive step when switching presentation styles.
